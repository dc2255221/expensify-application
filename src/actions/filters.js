// store action creators for filters


// SET_TEXT_FILTER Action Creator
export const setTextFilter = (text = '') => ({
    type: 'SET_TEXT_FILTER',
    text
});

// SORT_BY_DATE Action Creator
export const sortByDate = () => ({
    type: 'SORT_BY_DATE'
});

// SORT_BY_AMOUNT Action Creator
export const sortByAmount = () => ({
    type: 'SORT_BY_AMOUNT'
});

// SET_START_DATE Action Creator
export const setStartDate = (startDate) => ({
    type: 'SET_START_DATE',
    startDate
});

// SET_END_DATE Action Creator 
export const setEndDate = (endDate) => ({
    type: 'SET_END_DATE',
    endDate
});