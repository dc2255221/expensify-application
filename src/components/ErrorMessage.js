import React from 'react';
import { connect } from 'react-redux';

export const ErrorMessage = (props) => (
    <h3> {props.message} </h3>
);

const mapToStateToProps = (state) => (
    {
        message: state.error.message
    }
);

export default connect(mapToStateToProps)(ErrorMessage);
