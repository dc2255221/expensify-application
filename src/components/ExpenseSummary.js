import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectTotalExpenses from '../selectors/expenses-total';

export const ExpenseSummary = (props) => {
    const expensesTotal = numeral(selectTotalExpenses(props.expenses) / 100).format('$0,0.00');
    const expensesCount = props.expenses.length;
    return (
        <div>
            <h1> Viewing {expensesCount} {expensesCount === 1 ? 'expense' : 'expenses'} totalling {expensesTotal} </h1>
        </div>
    );
};

const mapToStateToProps = (state) => (
    {
        expenses: selectExpenses(state.expenses, state.filters)
    }
);

export default connect(mapToStateToProps)(ExpenseSummary);
