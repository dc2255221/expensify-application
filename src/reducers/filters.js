import moment from 'moment';
// Filters Reducer
const filtersReducerDefaultState = {
    category: '',
    text: '',
    sortBy: 'date',
    startDate: moment().startOf('month'),
    endDate: moment().endOf('month')
};

export default (state = filtersReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_CATEGORY':
            return {
                ...state,
                category: action.category
            }
        case 'RESET_CATEGORY':
            return {
                ...state,
                category: ''
            }
        case 'SET_TEXT_FILTER': 
            return {
                ...state,
                text: action.text
            };
        case 'SORT_BY_DATE':
            return {
                ...state,
                sortBy: 'date'
            };
        case 'SORT_BY_AMOUNT':
            return {
                ...state,
                sortBy: 'amount'
            };
        case 'SET_START_DATE':
            return {
                ...state,
                startDate: action.startDate
            };
        case 'SET_END_DATE':
            return {
                ...state,
                endDate: action.endDate
            };
        default: 
            return state;
    }
}
