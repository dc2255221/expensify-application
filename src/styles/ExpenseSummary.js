import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Colors, FontSize, Spacing } from '../constants';

export const PageHeader = styled.div`
    background: #f7f7f7;
    margin-bottom: ${Spacing.l_size};
    padding: ${Spacing.m_size} 0;
`;

export const ContentContainer = styled.div`
    margin: 0 auto;
    max-width: 80rem;
    padding: 0 ${Spacing.m_size};
`;

export const StyledH1 = styled.h1`
    font-weight: 300;
    margin: ${Spacing.s_size} 0;
`;

export const StyledSpan = styled.span`
    font-weight: 700;
`;

export const StyledLink = styled(Link)`
    background: ${Colors.dark_blue};
    border: none;
    color: white;
    display: inline-block;
    fondt-size: ${FontSize.large};
    font-weight: 300;
    line-height: 1;
    padding: ${Spacing.s_size};
    text-decoration: none;
`;

export const StyledDiv = styled.div`
    margin-top: ${Spacing.m_size};
`;
