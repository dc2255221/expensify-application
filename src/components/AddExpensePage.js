// The AddExpensePage will dispatch the given action to the redux store
// We need to connect component to store so it can dispatch
// Don't need anything from state so leave first parentheses empty but pass component to second one
// Now, we have access to props.dispatch
// We can call it within our onSubmit eventlistener
// We need addExpense action generator to create the action that we are going to dispatch
// When we did dispatch, we added the data to the redux store
// Now, we need to automatically redirect to Dashboard page onSubmit of form
// Components that we render inside of React.router gets access to many special props
// On history prop, we have a push method which is how we can programatically change pages. The method takes in the string path to Dashboard
// Using it will switch you over as if you clicked Dashboard link through Browser Routing. Hence, no full page refresh

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  };
};

// way to return dispatcher function, allowing you to abstract them from the components themselves
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);


