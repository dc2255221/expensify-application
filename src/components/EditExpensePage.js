// We need to connect this component to redux store so we can dispatch action 
// Search the array of expenses for the expense with given id 
// We do need to set up a mapToStateToProps function bc we need to give component access to the expense object
// We will be able to access the state which is great bc that is where expenses array lives
// But, we also need to pass in props as second argument bc we are searching for props.match.params.id
// We will add one new prop expense and will use current props to search expenses array
// Find allows us to search through an array looking for a single item. We determine whether item is found by returning true from the callback
// Now, we will render an instance of ExpenseForm which takes in an onSubmit prop
// We need to make sure data for the given expense is populated within the form
// We do this by passing down the expense as props to ExpenseForm and having it set the values to state 
// If no expense is passed down (i.e. the case of AddExpenseForm), set default values to state 
// Dispatch the action to edit the expense onSubmit of form
// Redirect to the dashboard using history's push method
// Move the removeExpense button from dashboard to EditExpensePage

// React router will render our connected higher order component 
// The higher order component passes the props through and also allows us to add some new ones within the returned object of mapToStateToProps function 

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/'); 
  }
  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  render() {
    return (
        <div>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button onClick={this.onRemove}> Remove </button> 
        </div>
    );
  };
};

const mapToStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense( id, expense )),
  removeExpense: (expenseData) => dispatch(removeExpense( expenseData ))
});

export default connect(mapToStateToProps, mapDispatchToProps)(EditExpensePage);