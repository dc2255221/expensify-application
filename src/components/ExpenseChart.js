import React from 'react';
import { connect } from 'react-redux';
import { Doughnut } from 'react-chartjs-2';
import selectExpenses from '../selectors/expenses';
import selectTotalByCategory from '../selectors/expenses-category';
import { ChartContainer } from '../styles/ExpenseChart';
import { CatColors } from '../constants';

export const ExpenseChart = ({ expenses }) => {
    const data = {
        datasets: [
            {
                data: [
                    selectTotalByCategory(expenses, 'home-utilities'),
                    selectTotalByCategory(expenses, 'transportation'),
                    selectTotalByCategory(expenses, 'groceries'),
                    selectTotalByCategory(expenses, 'personal-family-care'),
                    selectTotalByCategory(expenses, 'health'),
                    selectTotalByCategory(expenses, 'insurance'),
                    selectTotalByCategory(expenses, 'restaurants-dining'),
                    selectTotalByCategory(expenses, 'shopping-entertainment'),
                    selectTotalByCategory(expenses, 'travel'),
                    selectTotalByCategory(expenses, 'cash-checks-misc'),
                    selectTotalByCategory(expenses, 'giving'),
                    selectTotalByCategory(expenses, 'business-expenses'),
                    selectTotalByCategory(expenses, 'education'),
                    selectTotalByCategory(expenses, 'finance'),
                    selectTotalByCategory(expenses, 'uncategorized')
                ],
                backgroundColor: [
                    CatColors.light_green,
                    CatColors.med_light_green,
                    CatColors.green,
                    CatColors.med_light_green,
                    CatColors.dark_green,
                    CatColors.blue,
                    CatColors.yellow,
                    CatColors.light_orange,
                    CatColors.orange,
                    CatColors.dark_orange,
                    CatColors.pink,
                    CatColors.dark_pink,
                    CatColors.purple,
                    CatColors.dark_purple,
                    CatColors.grey
                ],
                legend: false
            }
        ],
        labels: [
            'Home and Utilities',
            'Transportation',
            'Groceries',
            'Personal and Family Care',
            'Health',
            'Insurance',
            'Restaurants and Dining',
            'Shopping and Entertainment',
            'Travel',
            'Cash, Checks, and Misc',
            'Giving',
            'Business Expenses',
            'Education',
            'Finance',
            'uncategorized'
        ]
    }
    return (
            !!expenses.length &&
            <ChartContainer>
                <Doughnut data={data}/>
            </ChartContainer> 
    )
};

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseChart);