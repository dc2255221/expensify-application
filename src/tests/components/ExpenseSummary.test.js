import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import getVisibleExpenses from '../../selectors/expenses';
import expenses from '../fixtures/expenses';
import {altFilters2} from '../fixtures/filters';

test('should correctly show total for one expense', () => {
    const wrapper = shallow(<ExpenseSummary expenses={[expenses[0]]}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should correctly show total for multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenses={expenses}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should correctly show total for filtered expenses', () => {
    const filteredExpenses = getVisibleExpenses(expenses, altFilters2);
    const wrapper = shallow(<ExpenseSummary expenses={filteredExpenses}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});
