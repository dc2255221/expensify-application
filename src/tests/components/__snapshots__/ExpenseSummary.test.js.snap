// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly show total for filtered expenses 1`] = `
<d>
  <h1>
     Viewing 
    2
     
    expenses
     totalling 
    $1,140.00
     
  </h1>
</d>
`;

exports[`should correctly show total for multiple expenses 1`] = `
<d>
  <h1>
     Viewing 
    3
     
    expenses
     totalling 
    $1,141.95
     
  </h1>
</d>
`;

exports[`should correctly show total for one expense 1`] = `
<d>
  <h1>
     Viewing 
    1
     
    expense
     totalling 
    $1.95
     
  </h1>
</d>
`;
